data_structures/binary_search_tree.ts

```diff
@@ -0,0 +1,199 @@
+  constructor(
+    public data: T,
+    public leftChild?: TreeNode<T>,
+    public rightChild?: TreeNode<T>,
+  ) {}
+}
+
+/**
+ * An implementation of a binary search tree.
+ *
+ * A binary tree is a tree with only two children per node. A binary search tree on top sorts the children according
+ * to following rules:
+ * - left child < parent node
+ * - right child > parent node
+ * - all children on the left side < root node
+ * - all children on the right side > root node
+ *
+ * For profound information about trees
+ * @see https://www.geeksforgeeks.org/introduction-to-tree-data-structure-and-algorithm-tutorials/
+ *
+ * @template T The data type of the values in the binary tree.
+ */
+export class BinarySearchTree<T> {
+  rootNode?: TreeNode<T>;
+
+  /**
+   * Instantiates the binary search tree.
+   *
+   * @param rootNode The root node.
+   */
+  constructor() {
+    this.rootNode = undefined;
+  }
+
+  /**
+   * Checks, if the binary search tree is empty, i. e. has no root node.
+   *
+   * @returns Whether the binary search tree is empty.
+   */
+  isEmpty(): boolean {
+    return this.rootNode === undefined;
+  }
+
+  /**
+   * Searches for data in the current binary search tree.
+   *
+   * @param data The data to search for.
+   */
+  search(data: T): TreeNode<T> {
+    if (!this.rootNode) {
+      throw new Error('No root node defined.');
+    }
+
+    let currentNode = this.rootNode;
+    while (currentNode.data !== data) {
+      if (data > currentNode.data) {
+        if (!currentNode.rightChild) {
+          throw new Error('No such element found.');
+        }
+
+        currentNode = currentNode.rightChild;
+      } else {
+        if (!currentNode.leftChild) {
+          throw new Error('No such element found.');
+        }
+
+        currentNode = currentNode.leftChild;
+      }
+    }
+
+    return currentNode;
+  }
+
+  /**
+   * Inserts the given data into the binary search tree.
+   *
+   * @param data The data to be stored in the binary search tree.
+   * @returns
+   */
+  insert(data: T): TreeNode<T> {
+    if (!this.rootNode) {
+      this.rootNode = new TreeNode<T>(data);
+      return this.rootNode;
+    }
+
+    let currentNode: TreeNode<T> = this.rootNode;
+    while (true) {
+      if (data > currentNode.data) {
+        if (currentNode.rightChild) {
+          currentNode = currentNode.rightChild;
+        } else {
+          currentNode.rightChild = new TreeNode<T>(data);
+          return currentNode.rightChild;
+        }
+      } else {
+        if (currentNode.leftChild) {
+          currentNode = currentNode.leftChild;
+        } else {
+          currentNode.leftChild = new TreeNode<T>(data);
+          return currentNode.leftChild;
+        }
+      }
+    }
+  }
+
+  /**
+   * Finds the minimum value of the binary search tree.
+   *
+   * @returns The minimum value of the binary search tree
+   */
+  findMin(): T {
+    if (!this.rootNode) {
+      throw new Error('No root node defined');
```

Comment:
I'd prefer a descriptive error message of `empty set` which abstracts away from the implementation details

Reply:
I might have misunderstood this, but I changed it so `Empty tree` as this is an implementation of a tree and not of a set. Feel free to correct me, if I did not understand you correctly. :)

Reply:
Binary search trees are used to implement a sorted set data structure, just like linked lists may be used to implement a stack.

Reply:
Alright, thank you for the clarification.